### PR TITLE
[Proposal] POEM: provide array result for all runtime images

### PR DESCRIPTION
**Pull request by ningyougang**
## Description
This is the POEM documentation for new feature providing `support array result for all runtime images`
I refer to this guide document for POEM: https://github.com/apache/openwhisk/blob/master/proposals/README.md

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Scheduler
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [x] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md#coding-standards) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by ningyougang** [comment]
Hi，guys.
   Regarding `prvoide array result`, for common action or sequence action, i have implemented in my local, e.g. for this example: https://github.com/apache/openwhisk/blob/master/docs/actions.md#creating-action-sequences

I updated the split.js and sort.js like below and invoked successully using my local openwhisk codes
```split.js
function main(msg) {
   var separator = msg.separator || /\r?\n/;
   var payload = msg.payload.toString();
   var lines = payload.split(separator);
   return lines;
}
```
And
```sort.js
function main(msg) {
   var lines = msg || [];
   console.log('sort before: ' + lines);
   lines.sort();
   return lines;
}

```

File: proposals/POEM-support-array-result.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+For sequence action, need to support as well.
+
+# Status
+* Current state: In-progress
+* Author(s): @ningyougang
+
+# Summary and Motivation
+
+This POEM proposes a new feature that allows user to write their own action which supports array result.
+So the result will support object and array both in future.
```
Comment (style95) [inline proposals/POEM-support-array-result.md:45]:
```suggestion
So actions would be able to return a JSON object or an array.
```

File: proposals/POEM-support-array-result.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+# Proposed changes
+## Openwhisk main repo
+Make controller and invoker support array result both.
+
+## Runtime repos
+All runtime images should support array result. e.g.
```
Comment (style95) [inline proposals/POEM-support-array-result.md:52]:
```suggestion
All runtime images should support an array result. e.g.
```

File: proposals/POEM-support-array-result.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+* Author(s): @ningyougang
+
+# Summary and Motivation
+
+This POEM proposes a new feature that allows user to write their own action which supports array result.
+So the result will support object and array both in future.
+
+# Proposed changes
+## Openwhisk main repo
+Make controller and invoker support array result both.
```
Comment (style95) [inline proposals/POEM-support-array-result.md:49]:
```suggestion
Make controller and invoker support both a JSON object and an array result.
```

File: proposals/POEM-support-array-result.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+* docker
+* ruby
+* dotnet
+* rust
+* swift 
+* deno
+* ballerina
+
+## Openwhisk-cli repo
+* When use wsk to execute action, need to support parse array result.
```
Comment (style95) [inline proposals/POEM-support-array-result.md:69]:
```suggestion
* The `wsk` CLI needs to support parsing an array result when executing actions.
```

File: proposals/POEM-support-array-result.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+* When use wsk to get activation, need to support parse array result as well.
```
Comment (style95) [inline proposals/POEM-support-array-result.md:70]:
```suggestion
* The `wsk` CLI needs to support parsing an array result when getting activations.
```

File: proposals/POEM-support-array-result.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+# Title
+
+Currently, openwhisk supports return json object only, e.g.
```
[the remark this answers is not in the document]
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+```shell
+# wsk action invoke hello -r
+{
+    "greeting": "Hello stranger!"
+}
+```
+It is necessary to support return array for common action, e.g.
```
[the remark this answers is not in the document]
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+```shell
+# wsk action invoke hello-array -r
+[
+    "a",
+    "b"
+]
+```
+For sequence action, need to support as well.
```
[the remark this answers is not in the document]
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+# Status
+* Current state: In-progress
+* Author(s): @ningyougang
+
+# Summary and Motivation
+
+This POEM proposes a new feature that allows user to write their own action which supports array result.
```
[the remark this answers is not in the document]
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md (hunk 1 above)
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md (hunk 2 above)
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md (hunk 4 above)
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

File: proposals/POEM-support-array-result.md (hunk 5 above)
Comment (ningyougang) [reply proposals/POEM-support-array-result.md]: Updated accordingly

**Review by style95 — APPROVED** [review]

**Comment by ningyougang** [comment]
Have any more comment?

**Pull request merged**